src/consensus/vote_parser.zig

```diff
@@ -0,0 +1,286 @@
+        .tower_sync_switch => |tss| .{
+            .{ .tower_sync = tss.tower_sync },
+            tss.hash,
+        },
+
+        .authorize,
+        .authorize_checked,
+        .authorize_with_seed,
+        .authorize_checked_with_seed,
+        .initialize_account,
+        .update_commission,
+        .update_validator_identity,
+        .withdraw,
+        => null,
+    };
+}
+
+test testParseVoteTransaction {
+    var prng = std.Random.DefaultPrng.init(42);
+    const random = prng.random();
+    try testParseVoteTransaction(null, random);
+    try testParseVoteTransaction(Hash.generateSha256(&[_]u8{42}), random);
+}
+
+fn testParseVoteTransaction(input_hash: ?Hash, random: std.Random) !void {
+    const allocator = std.testing.allocator;
+
+    const node_keypair = try randomKeyPair(random);
+    const auth_voter_keypair = try randomKeyPair(random);
+    const vote_keypair = try randomKeyPair(random);
+
+    const vote_key = Pubkey.fromPublicKey(&vote_keypair.public_key);
+
+    {
+        const bank_hash = Hash.ZEROES;
+        const vote_tx = try newVoteTransaction(
+            allocator,
+            &.{42},
+            bank_hash,
+            Hash.ZEROES,
+            node_keypair,
+            vote_key,
+            auth_voter_keypair,
+            input_hash,
+        );
+        defer vote_tx.deinit(allocator);
+
+        const maybe_parsed_tx = try parseVoteTransaction(allocator, vote_tx);
+        defer if (maybe_parsed_tx) |parsed_tx| parsed_tx.deinit(allocator);
+
+        try std.testing.expectEqualDeep(ParsedVote{
+            .key = vote_key,
+            .vote = .{ .vote = .{
+                .slots = &.{42},
+                .hash = bank_hash,
+                .timestamp = null,
+            } },
+            .switch_proof_hash = input_hash,
+            .signature = vote_tx.signatures[0],
+        }, maybe_parsed_tx);
+    }
+
+    // Test bad program id fails
+    var vote_ix = try vote_instruction.createVote(
+        allocator,
+        vote_key,
+        Pubkey.fromPublicKey(&auth_voter_keypair.public_key),
+        .{ .vote = .{
+            .slots = &.{ 1, 2 },
+            .hash = Hash.ZEROES,
+            .timestamp = null,
+        } },
+    );
+    defer vote_ix.deinit(allocator);
+    vote_ix.program_id = Pubkey.ZEROES;
+
+    const vote_tx = blk: {
+        const vote_tx_msg: TransactionMessage = try .initCompile(
+            allocator,
+            &.{vote_ix},
+            Pubkey.fromPublicKey(&node_keypair.public_key),
+            Hash.ZEROES,
+            null,
+        );
+        errdefer vote_tx_msg.deinit(allocator);
+        break :blk try Transaction.initOwnedMsgWithSigningKeypairs(
+            allocator,
+            .legacy,
+            vote_tx_msg,
+            &.{},
+        );
+    };
+    defer vote_tx.deinit(allocator);
+    try std.testing.expectEqual(null, parseVoteTransaction(allocator, vote_tx));
+}
+
+/// Reimplemented locally from Vote program.
+fn newVoteTransaction(
```

Comment:
Nit: if these functions are meant only for testing, it would be nice to have `builtin.is_test` guard for them.